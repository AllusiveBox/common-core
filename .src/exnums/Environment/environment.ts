import { EXnum } from "../EXnum";
import { EXnumable } from "../../types";

/**
 *
 * EXnum class used to represent application environments that an application can run in.
 *
 * @class Environment
 * @extends EXnum
 * @beta
 * @since Version 0.2.0
 *
 */
export default class Environment extends EXnum {

    /**
     *
     * Map that contains data for converting an Environment to an EXnumable value.
     *
     * @type {Map<Environment, string>}
     * @protected
     * @since Version 0.2.0
     *
     */
    protected static EXNUM_TO_VALUE_MAP: Map<Environment, string> = new Map<Environment, string>();

    /**
     *
     * Map that contains data for converting an EnvironmentEXnumable to a EXnum.
     *
     * @type {Map<string, Environment>}
     * @protected
     * @since Version 0.2.0
     *
     */
    protected static VALUE_TO_EXNUM_MAP: Map<string, Environment> = new Map<string, Environment>();

    /**
     *
     * Environment EXnum used to indicate a production environment.
     *
     * @type {Environment}
     * @public
     * @static
     * @since Version 0.2.0
     *
     */
    public static PRODUCTION: Environment = new Environment("PRODUCTION");

    /**
     *
     * Environment EXnum used to indicate a development environment that is <em>not</em> a local environment.
     *
     * @type {Environment}
     * @public
     * @static
     * @since Version 0.2.0
     *
     */
    public static DEVELOPMENT: Environment = new Environment("DEVELOPMENT");

    /**
     *
     * Environment EXnum used to indicate a development environment that is <em>local only</em>.
     *
     * @type {Environment}
     * @public
     * @static
     * @since Version 0.2.0
     *
     */
    public static LOCAL: Environment = new Environment("LOCAL");

    /**
     *
     * Environment EXnum used to indicate a test environment.
     *
     * @type {Environment}
     * @public
     * @static
     * @since Version 0.2.0
     *
     */
    public static TEST: Environment = new Environment("TEST");

    /**
     *
     * Environment EXnum used to indicate an unknown environment.
     *
     * @type {Environment}
     * @public
     * @static
     * @since Version 0.2.0
     *
     */
    public static UNKNOWN: Environment = new Environment("UNKNOWN");

    /**
     *
     * An array of Environments that the Environment EXnum supports.
     *
     * @type {Array<EXnum>}
     * @protected
     * @static
     * @readonly
     * @since Version 0.2.0
     *
     */
    protected static readonly SUPPORTED_EXNUMS: Array<EXnum> = [
        Environment.PRODUCTION,
        Environment.DEVELOPMENT,
        Environment.LOCAL,
        Environment.TEST,
        Environment.UNKNOWN
    ];

    /**
     *
     * An array of Environment EXnums that should support debug level logging.
     *
     * @type {Array<Environment>}
     * @protected
     * @static
     * @readonly
     *
     */
    protected static readonly SUPPORTS_DEBUG: Array<Environment> = [
        Environment.LOCAL,
        Environment.TEST,
        Environment.UNKNOWN
    ];

    /**
     *
     * The type for the Environment EXNum, at a static level.
     *
     * @type {string}
     * @protected
     * @static
     * @readonly
     * @since Version 0.2.0
     *
     */
    protected static readonly TYPE: string = "Environment";

    /**
     *
     * Gets the corresponding Environment value associated with the supplied EXnumable value. If no associated
     * Environment EXnum can be found, {@link Environment.UNKNOWN} is returned.
     * <br />
     * Alias for the {@link EXnum.getEXnum} method with the `errorIfNotFound` flag set to false.
     *
     * @param {EXnumable} value The value to get the Environment of.
     * @returns {Environment} The Environment associated with the EXnumable value.
     * @override
     * @since Version 0.2.0
     *
     */
    public static getEnvironment(
        value: string
    ): Environment {
        return this.getEXnum(
            value,
            false
        ) || Environment.UNKNOWN;
    }

    /**
     *
     * Gets the corresponding EXnumable value that maps to the provided Environment.
     *
     * @param {Environment} environment the environment to use to get the EXnumable value.
     * @returns {string} The string value associated with the provided Environment.
     * @override
     * @since Version 0.2.0
     *
     */
    public static override getValue(
        environment: Environment
    ): string {
        return this.EXNUM_TO_VALUE_MAP.get(environment) || "UNKNOWN";
    }

    /**
     *
     * Creates an instance of the Environment EXnum.
     *
     * @param {EXnumable} code The code that represents the Environment EXnum.
     * @private
     * @constructor
     * @since Version 0.2.0
     *
     */
    private constructor(
        code: EXnumable
    ) {
        super(
            code,
            "Environment"
        );
    }

    /**
     *
     * Sets the Environment EXnum maps with the provided value.
     *
     * @param {string} value The EXnumable value to map to the EXnum.
     * @returns {void}
     * @protected
     * @since Version 0.2.0
     *
     */
    protected setEXnumMap(
        value: string
    ): void {
        Environment.EXNUM_TO_VALUE_MAP.set(
            this,
            value
        );
        Environment.VALUE_TO_EXNUM_MAP.set(
            value,
            this
        );
    }

    /**
     *
     * Indicates if the Environment is {@link Environment.DEVELOPMENT}.
     *
     * @returns {boolean}
     * @since Version 0.2.0
     *
     */
    public isDevelopment(): boolean {
        return this === Environment.DEVELOPMENT;
    }

    /**
     *
     * Indicates if the Environment is {@link Environment.LOCAL}
     *
     * @returns {boolean}
     * @since Version 0.2.0
     *
     */
    public isLocal(): boolean {
        return this === Environment.LOCAL;
    }

    /**
     *
     * Indicates if the Environment is {@link Environment.PRODUCTION}
     *
     * @returns {boolean}
     * @since Version 0.2.0
     *
     */
    public isProduction(): boolean {
        return this === Environment.PRODUCTION;
    }

    /**
     *
     * Indicates if the Environment is {@link Environment.TEST}
     *
     * @returns {boolean}
     * @since Version 0.2.0
     *
     */
    public isTest(): boolean {
        return this === Environment.TEST;
    }

    /**
     *
     * Indicates if the Environment is {@link Environment.UNKNOWN}
     *
     * @returns {boolean}
     * @since Version 0.2.0
     *
     */
    public isUnknown(): boolean {
        return this === Environment.UNKNOWN;
    }

    /**
     *
     * Indicates if the Environment is configured to support debug logging.
     *
     * @returns {boolean}
     * @since Version 0.2.0
     *
     */
    public supportsDebug(): boolean {
        return Environment.SUPPORTS_DEBUG.includes(this);
    }

}
