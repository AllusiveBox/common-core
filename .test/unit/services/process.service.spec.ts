import "jest-extended";
import Environment from "../../../.src/znums/environment";

/*
 We need to mock the Environment getEnvironment method before importing the ProcessService, so that the static block
 is able to be manipulated during tests. Setting it to always return Unknown allows us the ability to check what
 was sent, and calculate what would have been returned, if not mocked.
 */
const mockEnvironmentSpy: jest.SpyInstance = jest.spyOn(
    Environment,
    "getEnvironment"
)
    .mockImplementation(() => {
        return Environment.UNKNOWN;
    });

import ProcessService from "../../../.src/services/process.service";
import LoggerServiceFake from "../../fakes/logger.service.fake";

/**
 *
 * The value passed to the `process.chdir` method during the tests.
 *
 * @type {string}
 *
 */
let chDirArg: string;

/**
 *
 * A logger used during the tests.
 *
 * @type {LoggerServiceFake}
 *
 */
let logger: LoggerServiceFake;

/**
 *
 * The service used during the tests.
 *
 * @type {ProcessService}
 *
 */
let service: ProcessService;

/**
 *
 * Test Suite for the ProcessService class.
 *
 * @group unit
 * @group services
 *
 */
describe(
    "ProcessService Unit Test Suite",
    () => {

        beforeEach(() => {
            jest.spyOn(
                process,
                "cwd"
            )
                .mockReturnValue("foo/bar");

            jest.spyOn(
                process,
                "chdir"
            )
                .mockImplementationOnce((arg: string) => {
                    chDirArg = arg;

                    const newDir: string = chDirArg === "../" ? "foo" : chDirArg;
                    jest.spyOn(
                        process,
                        "cwd"
                    )
                        .mockReturnValue(newDir);
                })
                .mockImplementationOnce(() => {
                    throw new Error("Error - chdir should only be called once");
                });
            mockEnvironmentSpy.mockImplementation(() => {
                return Environment.UNKNOWN;
            });

            logger = new LoggerServiceFake();
            service = new ProcessService(logger);
        });

        afterEach(() => {
            jest.resetAllMocks();
        });

        afterAll(() => {
            jest.restoreAllMocks();
        });

        test(
            "the static values initialize correctly",
            () => {
                expect(ProcessService.environment)
                    .toBeInstanceOf(Environment);
                // This will be false, as the getEnvironment method is mocked to always return Unknown.
                expect(ProcessService.environment.isTest())
                    .toBeFalse();

                expect(mockEnvironmentSpy)
                    .toHaveBeenCalledOnce();
                expect(mockEnvironmentSpy.mock.calls[0][0])
                    .toBe("test");
            }
        );

        test(
            "that the service initializes correctly",
            () => {
                expect(service.currentDirectory)
                    .toBe("foo/bar");
                expect(service.logger)
                    .toBeInstanceOf(LoggerServiceFake);

                // This will be false, as the getEnvironment method is mocked to always return Unknown.
                expect(service.isTest())
                    .toBeFalse();

                expect(mockEnvironmentSpy)
                    .toHaveBeenCalledOnce();
                expect(mockEnvironmentSpy.mock.calls[0][0])
                    .toBe("test");
            }
        );

        test(
            "that back correctly updates the current directory",
            () => {
                // Expect value to be unchanged; Base case
                expect(service.currentDirectory)
                    .toBe("foo/bar");

                // Update directory
                service.back();

                expect(service.currentDirectory)
                    .toBe("foo");
                expect(chDirArg)
                    .toBe("../");
            }
        );

        test(
            "that change correctly updates the current directory",
            () => {
                // Expect value to be unchanged; Base case
                expect(service.currentDirectory)
                    .toBe("foo/bar");

                // Update directory
                service.change("bar");

                expect(service.currentDirectory)
                    .toBe("bar");
                expect(chDirArg)
                    .toBe("bar");
            }
        );

        test(
            "that getEnvironment parses the process.env.NODE_ENV value if the stored Environment is Unknown",
            () => {
                expect(service.getEnvironment())
                    .toBe(Environment.UNKNOWN);

                expect(mockEnvironmentSpy)
                    .toHaveBeenCalledOnce();
                expect(mockEnvironmentSpy.mock.calls[0][0])
                    .toBe("test");
            }
        );

        test(
            "that getEnvironment correctly updates the internal environment value if the process.env.NODE_ENV value" +
            " is anything other than Unknown",
            () => {
                // Setup
                mockEnvironmentSpy.mockImplementationOnce((arg) => {
                    if (arg === "test") {
                        return Environment.TEST;
                    }

                    throw new Error(`Unrecognized value: ${arg}`);
                })
                    .mockImplementationOnce(() => {
                        throw new Error("Error - getEnvironment should only be called once");
                    });

                expect(service.getEnvironment())
                    .toBe(Environment.TEST);

                expect(mockEnvironmentSpy)
                    .toHaveBeenCalledOnce();
                expect(mockEnvironmentSpy.mock.calls[0][0])
                    .toBe("test");
            }
        );

        test(
            "that getEnvironment returns the stored environment, when it is anything other than Unknown",
            () => {
                // Setup
                mockEnvironmentSpy.mockImplementationOnce((arg) => {
                    if (arg === "test") {
                        return Environment.TEST;
                    }

                    throw new Error(`Unrecognized value: ${arg}`);
                })
                    .mockImplementationOnce(() => {
                        throw new Error("Error - getEnvironment should only be called once");
                    });

                expect(service.getEnvironment())
                    .toBe(Environment.TEST);
            }
        );

    }
);