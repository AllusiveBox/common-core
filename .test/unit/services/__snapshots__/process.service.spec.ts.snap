// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessService Unit Test Suite that given a directory mismatch "change" throws a ProcessServiceError 1`] = `"ENOENT: no such file or directory, chdir 'C:\\Users\\megaz\\workspace' -> ''"`;
